juju/juju
# [JUJU-2310] Secret multi backends

Add support for multiple secret backends per model.

Previously, a model used just one secret backend. Now, the backend can be changed and new revisions are stored in the new backend. Older revisions are still fetched from the backend they were stored in.

A key change is to model external secret value references to include the backend id as well as revision id. This needed to be weaved through the code.

The provider interfaces are improved to take config structs directly instead of a model interface.
Operations like model cleanup now need to iterate through all configured backends to deal with revisions in any of them

Broken up into 4 commits - state, provider, uniter, facades.


## QA steps

add a model and deploy dummy-source and dummy-sink and relate
create a secret in dummy-source
check mongo that secret content is stored there
dummy-sink cannot read secret yet
grant access to dummy-sink, check secret-get

add vault as a backend
repeat the above steps, checking secret content is now in vault

show-secret --reveal on both secrets will print content
destroy model will cleanup secret in vault

repeat above on a k8s model, except out of the box, secrets are stored in k8s not mongo
